### ISSUE TITLE
Haddock runs out of memory when building documentation for GHC

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/haddock#1061]
On GHC commit [`6f07f82`](https://gitlab.haskell.org/ghc/ghc/commit/6f07f828e4f7a445fabd82dcb3fbf6edb2641369) (which is pinned against Haddock commit a31b562f3f3a33ae2b08f9173ea08506b357f3a5), building GHC with the `quick` build flavor and the Make-based build system causes Haddock to run out of memory before it can finish building the documentation for `ghc`-the-library:

```
Warning: 'Lift' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcInferRho' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'getInitialKind' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcImplicitTKBndrs' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'wanted' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcRnSrcDecls' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'KindRep' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
haddock: out of memory
compiler/ghc.mk:430: recipe for target 'compiler/stage2/doc/html/ghc/ghc.haddock' failed
make[1]: *** [compiler/stage2/doc/html/ghc/ghc.haddock] Error 251
Makefile:123: recipe for target 'all' failed
make: *** [all] Error 2
```

Some observations:

* It appears to run out of memory in the exact same spot every time I try this (right after `KindRep`).
* I haven't yet tried this with a build flavour besides `quick` or using the equivalent Hadrian configuration.

**Comment by harpocrates** [comment]
This is not good. I'll make time to investigate this in coming days.

Is the problem also coming up on the 8.8 branch?

**Comment by RyanGlScott** [comment]
The plot thickens: I can _not_ trigger this issue with `./hadrian/build.sh --flavour=quick -j4 docs-haddock`. Yikes.

I'll try the `ghc-8.8` branch next.

**Comment by harpocrates** [comment]
I can reproduce this on the 8.8 branch 😢. I wonder what change could've triggered this.

**Comment by RyanGlScott** [comment]
Urk, I think I ran into a separate bug when I tried to build the `ghc-8.8` branch with `happy-1.19.10`:

```
compiler/stage1/build/Parser.hs:1487:48: error:
    Not in scope: type variable ‘a’
     |
1487 | newtype HappyWrap211 = HappyWrap211 (([Located a],Bool))
     |                                                ^
```

(Curiously, I didn't encounter this problem on `master`.)

In any case, thank you for confirming that the problem also exists on `ghc-8.8`—that saves me some detective work.

**Comment by harpocrates** [comment]
@RyanGlScott That error worries me a bit. Am I to understand that `happy-1.19.10` no longer accepts polymorphic signatures!?

For `ghc-8.8`, the issue is simpler: we need to cherry-pick https://github.com/ghc/ghc/commit/071bef18ccafb134ab886da9e362689b9f092dd3.

**Comment by RyanGlScott** [comment]
> That error worries me a bit. Am I to understand that `happy-1.19.10` no longer accepts polymorphic signatures!?

I'm far from an expert on `happy`, but judging from simonmar/happy#134 (which is the primary change introduced in `happy-1.19.10`), it looks like polymorphic productions now require an explicit `forall`.

Also, good catch regarding ghc/ghc@071bef1. After cherry-picking that commit to the `ghc-8.8` branch, I can confirm that I experience the same out-of-memory issue with `make` (but again, not with Hadrian). This time, it runs out of memory in a different spot:

```
Warning: 'TcM' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcPluginInit' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcPluginStop' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'splitPiTys'' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'Rep1' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcInferRho' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'getInitialKind' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'tcImplicitTKBndrs' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'wanted' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
haddock: out of memory
compiler/ghc.mk:445: recipe for target 'compiler/stage2/doc/html/ghc/ghc.haddock' failed
make[1]: *** [compiler/stage2/doc/html/ghc/ghc.haddock] Error 251
Makefile:123: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by harpocrates** [comment]
I'm (slowly) bisecting with GHC's 20f8f2f98bd686b297e7290b2348ac7e937c1a5a as the good commit.

**Comment by harpocrates** [comment]
This looks like a regression in GHC's RTS. I've got a patch hopefully incoming...

**Comment by harpocrates** [comment]
Got a fix up at https://gitlab.haskell.org/ghc/ghc/merge_requests/942.

**Comment by harpocrates** [comment]
@RyanGlScott if you clean everything first, are you sure you can't replicate this bug with `./hadrian/build.sh --flavour=quick -j4 docs-haddock`? If it turns out you can, it sounds like there is also a Hadrian bug lurking about.

**Comment by RyanGlScott** [comment]
Wow, incredible sleuthing @harpocrates!

In response to https://github.com/haskell/haddock/issues/1061#issuecomment-492018699, I did a very aggressive clean (by which I mean `git clean -fdx`), followed by `./boot`, `./configure`, and `./hadrian/build.sh --flavour=quick -j4 docs-haddock`. I did get a failure when building the Haddocks although curiously, it was _not_ when building the Haddocks for `ghc`-the-libary, but rather `ghc-prim`:

```
| Run Haddock BuildPackage: libraries/ghc-prim/GHC/CString.hs (and 8 more) => _build/docs/html/libraries/ghc-prim/ghc-prim.haddock
Error when running Shake build system:
  at want, called at src/Main.hs:75:30 in main:Main
* Depends on: docs-haddock
  at need, called at src/Rules/Documentation.hs:77:9 in main:Rules.Documentation
* Depends on: _build/docs/html/libraries/index.html
  at need, called at src/Rules/Documentation.hs:161:9 in main:Rules.Documentation
* Depends on: _build/docs/html/libraries/ghc-prim/ghc-prim.haddock
* Raised the exception:
user error (Development.Shake.cmd, system command failed
Command line: _build/stage1/bin/haddock --verbosity=0 -B_build/stage1/lib --lib=_build/stage1/lib --odir=_build/docs/html/libraries/ghc-prim --no-tmp-comp-dir --dump-interface=_build/docs/html/libraries/ghc-prim/ghc-prim.haddock --html --hyperlinked-source --hoogle --quickjump '--title=ghc-prim-0.6.1: GHC primitives' --prologue=_build/docs/html/libraries/ghc-prim/haddock-prologue.txt --optghc=-D__HADDOCK_VERSION__=2220 --optghc=-hisuf --optghc=dyn_hi --optghc=-osuf --optghc=dyn_o --optghc=-hcsuf --optghc=dyn_hc --optghc=-fPIC --optghc=-dynamic --optghc=-hide-all-packages --optghc=-no-user-package-db '--optghc=-this-unit-id ghc-prim-0.6.1' '--optghc=-package-id rts-1.0' --optghc=-i --optghc=-i_build/stage1/libraries/ghc-prim/build --optghc=-i_build/stage1/libraries/ghc-prim/build/autogen --optghc=-ilibraries/ghc-prim/. --optghc=-Iincludes --optghc=-I_build/generated --optghc=-I_build/stage1/libraries/ghc-prim/build --optghc=-I/home/rgscott/Software/ghc3/_build/stage1/lib/x86_64-linux-ghc-8.9.0.20190510/rts-1.0/include --optghc=-I_build/generated --optghc=-optc-I_build/generated --optghc=-optP-include --optghc=-optP_build/stage1/libraries/ghc-prim/build/autogen/cabal_macros.h --optghc=-outputdir --optghc=_build/stage1/libraries/ghc-prim/build --optghc=-this-unit-id --optghc=ghc-prim --optghc=-XHaskell2010 --optghc=-ghcversion-file=/home/rgscott/Software/ghc3/_build/generated/ghcversion.h --optghc=-Wno-deprecated-flags --optghc=-Wno-trustworthy-safe libraries/ghc-prim/GHC/CString.hs libraries/ghc-prim/GHC/Classes.hs libraries/ghc-prim/GHC/Debug.hs libraries/ghc-prim/GHC/IntWord64.hs libraries/ghc-prim/GHC/Magic.hs _build/stage1/libraries/ghc-prim/build/GHC/PrimopWrappers.hs libraries/ghc-prim/GHC/Tuple.hs libraries/ghc-prim/GHC/Types.hs _build/stage1/libraries/ghc-prim/build/GHC/Prim.hs +RTS -t_build/stage1/libraries/ghc-prim/build/haddock.t --machine-readable -RTS
Exit code: 1
Stderr:
)
```

That doesn't really tell me what went wrong, so I tried running the indicated command directly:

```
$ _build/stage1/bin/haddock --verbosity=0 -B_build/stage1/lib --lib=_build/stage1/lib --odir=_build/docs/html/libraries/ghc-prim --no-tmp-comp-dir --dump-interface=_build/docs/html/libraries/ghc-prim/ghc-prim.haddock --html --hyperlinked-source --hoogle --quickjump '--title=ghc-prim-0.6.1: GHC primitives' --prologue=_build/docs/html/libraries/ghc-prim/haddock-prologue.txt --optghc=-D__HADDOCK_VERSION__=2220 --optghc=-hisuf --optghc=dyn_hi --optghc=-osuf --optghc=dyn_o --optghc=-hcsuf --optghc=dyn_hc --optghc=-fPIC --optghc=-dynamic --optghc=-hide-all-packages --optghc=-no-user-package-db '--optghc=-this-unit-id ghc-prim-0.6.1' '--optghc=-package-id rts-1.0' --optghc=-i --optghc=-i_build/stage1/libraries/ghc-prim/build --optghc=-i_build/stage1/libraries/ghc-prim/build/autogen --optghc=-ilibraries/ghc-prim/. --optghc=-Iincludes --optghc=-I_build/generated --optghc=-I_build/stage1/libraries/ghc-prim/build --optghc=-I/home/rgscott/Software/ghc3/_build/stage1/lib/x86_64-linux-ghc-8.9.0.20190510/rts-1.0/include --optghc=-I_build/generated --optghc=-optc-I_build/generated --optghc=-optP-include --optghc=-optP_build/stage1/libraries/ghc-prim/build/autogen/cabal_macros.h --optghc=-outputdir --optghc=_build/stage1/libraries/ghc-prim/build --optghc=-this-unit-id --optghc=ghc-prim --optghc=-XHaskell2010 --optghc=-ghcversion-file=/home/rgscott/Software/ghc3/_build/generated/ghcversion.h --optghc=-Wno-deprecated-flags --optghc=-Wno-trustworthy-safe libraries/ghc-prim/GHC/CString.hs libraries/ghc-prim/GHC/Classes.hs libraries/ghc-prim/GHC/Debug.hs libraries/ghc-prim/GHC/IntWord64.hs libraries/ghc-prim/GHC/Magic.hs _build/stage1/libraries/ghc-prim/build/GHC/PrimopWrappers.hs libraries/ghc-prim/GHC/Tuple.hs libraries/ghc-prim/GHC/Types.hs _build/stage1/libraries/ghc-prim/build/GHC/Prim.hs +RTS -t_build/stage1/libraries/ghc-prim/build/haddock.t --machine-readable -RTS
haddock: internal error: _build/stage1/lib/llvm-targets: openFile: does not exist (No such file or directory)
```

Bah, something about `llvm-targets`? I'm starting to think that I ran into another issue... in any case, I still figured out how to trigger the elusive out-of-memory error on Hadrian, although this might be due to my own inexperience with how Hadrian works.

**Comment by harpocrates** [comment]
I’ll take a look at Hadrian too. I’m in the process of moving cities, so my spare dev time is limited.

The error you ran into about llvm-targets sounds like Hadrian is forgetting to do something. Did you perhaps not re-run boot or configure? If not, please open a bug! A likely workaround would be to do a build of the default target, then run ‘docs-haddock’ after that.

One other tidbit: Haddock run with Hadrian uses hyperlinked sources, so it is possible that the memory profile ends up being different than with Make. That said, I’d be surprised if hyperlinked sources turned out to be the cause of peak heap usage.

Cheers!

**Comment by RyanGlScott** [comment]
> Did you perhaps not re-run boot or configure? If not, please open a bug!

I definitely ran `./boot` and `./configure`, as mentioned in https://github.com/haskell/haddock/issues/1061#issuecomment-492027102. (I wouldn't even have made it to the Haddock phase if not, since Hadrian topples over pretty quickly if you forget to `./boot` or `./configure` beforehand.)

> A likely workaround would be to do a build of the default target, then run ‘docs-haddock’ after that.

Indeed, I ran `./hadrian/build.sh --flavour=quick -j4` after getting stuck (which built some extra things) followed by `./hadrian/build.sh --flavour=quick -j4 docs-haddock`, and that seemed to bypass the `llvm-targets` issue. At that point, Haddocks successfully built the documentation for everything without any out-of-memory issues.

**Comment by harpocrates** [comment]
Looks like the `llvm-targets` issue really is a Hadrian bug then. Probably just a matter of `need`-ing some extra set of dependencies somewhere.

The fact that Hadrian succeeds where Make fails is interesting. I know that Hadrian passes `-O2` when building Haddock (even in `quick` I think). Perhaps Make doesn't, leading to a more memory hungry Haddock executable, thereby triggering https://gitlab.haskell.org/ghc/ghc/merge_requests/942.

**Comment by RyanGlScott** [comment]
Now that https://gitlab.haskell.org/ghc/ghc/merge_requests/942 has landed on GHC's `master` branch, I've opened https://gitlab.haskell.org/ghc/ghc/merge_requests/1022 to backport it to the `ghc-8.8` branch.

**Comment by harpocrates** [comment]
Everything is fixed here. Technically, there never really was a Haddock issue in the first place either. 😄

**Comment by jwaldmann** [comment]
not sure whether it's the same issue: I am on a machine with 6 GB RAM. I can compile ghc-8.8.3 from source (with ghc-8.6.5) but I cannot compile ghc-8.10.1 from source (with neither 8.8.3 nor 8.6.5) because haddock runs out of memory.
